File: app/models/embedding.py
"""Pydantic models for embedding generation, progress, and status."""

from pydantic import BaseModel


class EmbeddingGenerateRequest(BaseModel):
    """Request body for triggering embedding generation."""

    model_name: str = "siglip-base"


class EmbeddingGenerateResponse(BaseModel):
    """Response returned when embedding generation is triggered."""

    dataset_id: str
    status: str
    message: str


class EmbeddingProgress(BaseModel):
    """Progress update for embedding generation (used in SSE streaming)."""

    status: str = "idle"
    processed: int = 0
    total: int = 0
    message: str = ""


class EmbeddingStatus(BaseModel):
    """Current embedding status for a dataset."""

    dataset_id: str
    has_embeddings: bool
    embedding_count: int
    model_name: str | None = None
    has_reduction: bool


class ReductionProgress(BaseModel):
    """Progress update for dimensionality reduction (used in SSE streaming).

    Statuses:
    - ``idle``: No reduction has been started.
    - ``running``: Loading embeddings from the database.
    - ``fitting``: UMAP fit_transform in progress (can take 10-60s).
    - ``complete``: Reduction finished, 2D coordinates stored.
    - ``error``: An error occurred during reduction.
    """

    status: str = "idle"
    message: str = ""


class EmbeddingPoint(BaseModel):
    """A single 2D point for the scatter-plot visualization.

    Returned by GET /coordinates, one per embedded sample.
    """

    sample_id: str
    x: float
    y: float
    file_name: str
    thumbnail_path: str | None

File: frontend/src/types/embedding.ts
/**
 * TypeScript types matching backend embedding Pydantic models
 * from app/models/embedding.py.
 *
 * These types mirror the JSON shapes returned by the embeddings API
 * (GET /status, GET /coordinates, SSE progress events).
 */

/** A single 2D point for the scatter-plot visualization. */
export interface EmbeddingPoint {
  sampleId: string;
  x: number;
  y: number;
  fileName: string;
  thumbnailPath: string | null;
  gtLabel?: string | null;
  predLabel?: string | null;
}

/** Current embedding status for a dataset (GET /status). */
export interface EmbeddingStatus {
  dataset_id: string;
  has_embeddings: boolean;
  embedding_count: number;
  model_name: string | null;
  has_reduction: boolean;
}

/** Progress update from SSE streams (generation or reduction). */
export interface EmbeddingProgress {
  status: "idle" | "running" | "fitting" | "complete" | "error";
  processed?: number;
  total?: number;
  message: string;
}
